oac_summary.m4

```diff
@@ -0,0 +1,94 @@
+# shell-variable-ified representation of line.
+#
+# There are no restrictions on the contents of section and topic; they
+# can be variable references (although the use case for this is
+# dubious) and they can contain most ASCII characters (escape
+# characters excluded).  Note that some care must be taken with the
+# unique check and this liberal rule, as the unique check is after the
+# string has been run through AS_TR_SH.  Basically, any character that
+# is not legal in a shell variable name will be turned into an
+# underscore.  So the strings "Section_foo" and "Section-foo" would be
+# the same as far as the unique check is concerned.
+#
+# The input strings are evaluated during OAC_SUMMARY_ADD, not during
+# OAC_SUMMARY_PRINT.  This seems to meet the principle of least
+# astonishment.  A common pattern is to call
+# OAC_SUMMARY_ADD([Resource Type], [Component Name], [$results])
+# and then unset $results to avoid namespace pollution.  This will
+# work properly with the current behavior, but would result in odd
+# errors if we delayed evaluation.
+# -----------------------------------------------------------
+AC_DEFUN([OAC_SUMMARY_ADD],[
+    OAC_VAR_SCOPE_PUSH([oac_summary_line oac_summary_newline oac_summary_key])
+
+    # The end quote on the next line is intentional!
+    oac_summary_newline="
+"
+    oac_summary_line="$2: $3"
+    oac_summary_key="AS_TR_SH([$1])_AS_TR_SH([$2])"
+
+    # Use the section name variable as an indicator for whether or not
+    # the section has already been created.
+    AS_IF([AS_VAR_TEST_SET([oac_summary_section_]AS_TR_SH([$1])[_name])],
+          [],
+          [AS_VAR_SET([oac_summary_section_]AS_TR_SH([$1])[_name], ["$1"])
+           OAC_APPEND([oac_summary_sections], [AS_TR_SH([$1])])])
+
+    # Use the summary key as indicator if the section:topic has already
+    # been added to the results for the given section.
+    AS_IF([AS_VAR_TEST_SET([${oac_summary_key}])],
+          [],
+          [AS_VAR_SET([${oac_summary_key}], [1])
+           dnl this is a bit overcomplicated, but we are basically implementing
+           dnl a poor mans AS_VAR_APPEND with the ability to specify a separator,
+           dnl because we have a newline separator in the string.
+           AS_IF([AS_VAR_TEST_SET([oac_summary_section_]AS_TR_SH([$1])[_value])],
+                 [AS_VAR_APPEND([oac_summary_section_]AS_TR_SH([$1])[_value],
+                                ["${oac_summary_newline}${oac_summary_line}"])],
+                 [AS_VAR_SET([oac_summary_section_]AS_TR_SH([$1])[_value],
+                             ["${oac_summary_line}"])])])
+
+    OAC_VAR_SCOPE_POP
+])
+
+dnl If $1 is "stderr", then send the output to stderr.  Otherwise,
+dnl send the output to stdout.
+AC_DEFUN([OAC_SUMMARY_PRINT],[
+    OAC_VAR_SCOPE_PUSH([oac_summary_section oac_summary_section_name])
+    m4_define([oac_summary_print_fd], [m4_default(m4_if([$1], [stderr], [2]), [1])])
```

Comment:
Sorry, I didn't look at this closely enough.  I don't think this is right (or at least, it doesn't make sense).  If you want the valid values of the argument to be 'stderr' or not set, then test for those two cases.  I had provided code for the case where users could pass an arbitrary fd (which is why the use of the string stderr bugs me, because it's not an fd), in which case you need an else clause in your m4_if to pass $1.

As it stands now, passing anything other than stderr will result in the argument being ignored and printing to file descriptor 1.  I would say that we should either use the passed argument as an FD OR explicitly check for the three cases (not set, stderr, anything else) and abort if anything else is passed.

Finally, I thought you were going to use AS_MESSAGE_FD , not 1, as the default?